### PR TITLE
[CA-1312] A11y: Table roles

### PR DESCRIPTION
**Pull request by andrewmarcus**
This PR adds proper table rows to every type of table, to improve the accessibility of many pages within the app. This applies to the following table types:

* SimpleTable
* SimpleFlexTable
* FlexTable
* GridTable

### Table Names

* Each of the 4 table types now supports an optional `tableName` property, which is converted to an `aria-label` on the root table element. If not specified, it will be "data table" by default.

### Table Roles

* Each table has `role="table"` at the top level. This role does not imply that the table supports keyboard navigation. If we go back later and add keyboard navigation in order to make it easier to scroll through the virtualized DOM, then at that point these should be switched to `role="grid"` to imply that [keyboard navigation is possible](https://www.digitala11y.com/grid-role/).
* The header row has `role="row"` and each header column has `role="columnheader"`.
* Each data row has `role="row"` and each data cell has `role="cell"`. If the top-level is changed to `role="grid"` in the future to support arrow-key navigation, then each cell should be changed to `role="gridcell"`.

### Virtualized Tables

In FlexTable and GridTable, not all cells will be included in the DOM. To support this:
* Each cell now includes `aria-rowindex` and `aria-colindex` attributes to indicate where in the table these cells are supposed to be.
* Each of the 3 main table types now also includes consistent `rowIndex` and `columnIndex` properties when calling its respective `cellRenderer` functions. I didn't change existing behavior, but added new optional properties that could be used by table definitions if they wish. These two properties should now be presented consistently regardless of which table type you're using.

### Sorting

Table definitions currently declare `Sortable` headers within the output of the `headerRenderer()` function for each column. However, to properly use the `aria-sort` ARIA attribute, the attribute must be placed directly on the element with `role="columnheader"`, which is several levels higher in the DOM. To support this:
* The `Sortable` and `MiniSortable` renderers now produce columns which are keyboard interactive, and provide an ARIA description indicating that the column will resort itself when clicking on it.
* Each sortable table must now be passed the `sort` state object.
* Each column which supports sorting must now specify the sorting field name, the same way it is defined in the `Sortable` renderer.

_Note:_ I don't like the duplication. It's problematic for a developer to have to add each field name in two places -- in the column definition and again in the `Sortable` definition -- and to keep them in sync. I'd prefer to try to recursively find a `Sortable` object in the `children` array and grab the 'sort' and `field` properties from there. But so far I've been trying to avoid that under the impression that it's bad practice to try to grab attributes from inside hyperscript children arrays. I certainly haven't seen this done anywhere else.

Is there a better way to do this without adding duplicate work for developers?

### Page-Specific updates

I've added duplicate field names to support `aria-sort` to:

* UploadPreviewTable.js (/#upload)
* Environment.js (/#clusters)
* List.js (/#workspaces/list)
* DataTable.js (/#workspaces/data)
* SubmissionDetails.js (/#workspaces/:workspace/jobHistory
* WorkflowView.js (/#workspaces/:workspace/workflows
* List.js (/#workflows/list)

### CODE REVIEW

File: src/components/data/DataTable.js [hunk 1]
```diff
@@ -233,6 +235,7 @@ const DataTable = props => {
                   }
                 },
                 {
+                  field: 'name',
```
Comment (marctalbott) [inline src/components/data/DataTable.js:238]: Are the single quotes around `name` here intentional? If so, what's the difference between this field and the other fields where name isn't enclosed in single quotes?

File: src/components/table.js [hunk 2]
```diff
@@ -259,37 +277,52 @@ FlexTable.propTypes = {
...
+      role: 'row'
+    }, [
+      _.map(([i, { size, headerRenderer, field }]) => {
         return div({
           key: i,
-          style: { ...styles.flexCell(size), ...styles.header(i * 1, columns.length) }
-        }, [headerRenderer()])
+          style: { ...styles.flexCell(size), ...styles.header(i * 1, columns.length) },
+          role: 'columnheader',
+          'aria-sort': !sort || !field ? null : sort.field === field ? sort.direction === 'asc' ? 'ascending' : 'descending' : 'none'
```
Comment (marctalbott) [inline src/components/table.js]: Is there a way to extract this out into a variable? I see it is also used in other tables

File: src/components/table.js [hunk 3]
```diff
@@ -177,25 +178,33 @@ export const FlexTable = ({
     body.current.scrollToPosition({ scrollTop: initialY })
   })
 
-  return div([
+  return div({
+    role: 'table',
+    'aria-rowcount': rowCount,
```
Comment (marctalbott) [inline src/components/table.js]: Not sure if this is how it's supposed to be, but when I was testing this out locally, I found that this `rowCount` doesn't include the column headers, but the row indices do. Should those match up? It seems inconsistent to have a table report that there are 2 rows, but then have the index of the last row be 3

File: src/components/data/DataTable.js (hunk 1 above)
Comment (andrewmarcus) [reply src/components/data/DataTable.js:238]: No difference. I agree they might as well be consistent. Is there a preferred quote type (single vs double) in the repo?

File: src/components/table.js (hunk 2 above)
Comment (andrewmarcus) [reply src/components/table.js]: Hmm. I could calculate it each time within the renderer function and then just use the value here, if that would be preferred. I suppose I could also make this a shared function for all 3 tables, but it would basically be a one-liner.  Probably shouldn't memoize it, though, since the whole point of these virtual tables is to not have the entire thing in memory at a time...

What I would ideally prefer to do is extract it from the `Sortable` renderer nested inside the results from the `headerRenderer()` call. See comments above about avoiding duplication...  I'm just not sure whether it's good practice to introspect hyperscript objects. If you think that's ok, I can remove the `field` property here entirely.

File: src/components/table.js (hunk 3 above)
Comment (andrewmarcus) [reply src/components/table.js]: Good catch!

File: src/components/data/DataTable.js (hunk 1 above)
Comment (andrewmarcus) [reply src/components/data/DataTable.js:238]: Ah. Here the name of the field is `'name'` to match up with the `Sortable` definition attached to the column. In the situation below, `name` is a variable attached to each column created by the iterator

File: src/components/table.js (hunk 2 above)
Comment (s-rubenstein) [reply src/components/table.js]: I would probably extract it into a function, sort of like:
`(sort, field) => !sort || !field ? null : sort.field === field ? sort.direction === 'asc' ? 'ascending' : 'descending' : 'none'`, and then each place this is duplicated, we can just call that function. I think it could make things a little more readable.

File: src/components/data/DataTable.js (hunk 1 above)
Comment (andrewmarcus) [reply src/components/data/DataTable.js:238]: I agree this is confusing, but I just pulled up the column names that were already in the corresponding `Sortable` definitions.

File: src/components/table.js [hunk 4]
```diff
@@ -161,14 +161,18 @@ const styles = {
 // Note: We always need 1 extra row's worth of height for the table header row:
 export const tableHeight = ({ actualRows, maxRows, heightPerRow = 48 }) => (_.min([actualRows, maxRows]) + 1) * heightPerRow
 
+const ariaSort = (sort, field) => !sort || !field ? null :
+  sort.field !== field ? 'none' : sort.direction === 'asc' ? 'ascending' : 'descending'
+
```
Comment (andy7i) [inline src/components/table.js]: I had some trouble parsing this. Would it be possible to refactor this to make it more readable? Maybe not really possible but here are a couple suggestions:

* I think the ordering of negative case first and positive case last puts the most important piece when it might be easier if it came first. Not exactly the same but related: https://softwareengineering.stackexchange.com/questions/223086/should-if-else-statements-be-arranged-by-rareness-of-cases-or-difficulty-of-deal/223093
  * I'd be ok with keeping the current order if you feel strongly about it

* There are a lot of `!`s to keep track of. There's a comment about that in the link above
> Personally I would concentrate on avoiding not conditions. They have been proven to induce more cognitive load to understand than "positive" conditions and are detrimental to readability / understandability of code. I tend to use them in guard statements only

* the ternary chain is hard to follow. I think ternary operators are best for simple one-liners. Maybe it's also the fact that it's all on 2 lines.

Taking all of that into consideration, what do you think about this? I _think_ it should be functionally equivalent but can guarantee nothing!

```suggestion
const ariaSort = (sort, field) => {
  if (!!sort || !!field)
    if (field == sort.field)
      if (sort.direction === 'asc')
        'ascending'
      else 'descending'
    else 'none'
  else null
}
```

File: src/components/table.js [hunk 5]
```diff
@@ -212,20 +228,24 @@ export const FlexTable = ({
...
           style: { ...data.style, backgroundColor: 'white', display: 'flex' },
           hover: hoverHighlight ? { backgroundColor: colors.light(0.4) } : undefined
         }, [
           _.map(([i, { size, cellRenderer }]) => {
             return div({
               key: i,
               className: 'table-cell',
+              role: 'cell',
+              'aria-rowindex': data.rowIndex + 2, // The header row is 1
+              'aria-colindex': i + 1,
```
Comment (andy7i) [inline src/components/table.js]:
```suggestion
              'aria-colindex': i + 1, // aria-colindex must be ≥ 1 https://www.w3.org/TR/wai-aria-1.1/#aria-colindex
```

File: src/components/table.js [hunk 6]
```diff
@@ -259,37 +280,52 @@ FlexTable.propTypes = {
...
 
 /**
  * A basic table with a header and flexible column widths. Intended for small amounts of data,
  * since it does not provide scrolling. See FlexTable for prop types.
  */
-export const SimpleFlexTable = ({ columns, rowCount, noContentMessage, hoverHighlight }) => {
-  return h(Fragment, [
-    div({ style: { height: 48, display: 'flex' } }, [
-      _.map(([i, { size, headerRenderer }]) => {
+export const SimpleFlexTable = ({ columns, rowCount, noContentMessage, hoverHighlight, tableName = 'data table', sort = null }) => {
```
Comment (andy7i) [inline src/components/table.js:320]: I'm worried that it might be easy to forget to put a descriptive name in `tableName` if we provide a default value here.

File: src/pages/workspaces/workspace/workflows/WorkflowView.js [hunk 7]
```diff
@@ -108,9 +108,11 @@ const WorkflowIOTable = ({ which, inputsOutputs: data, config, errors, onChange,
   return h(SimpleFlexTable, {
     rowCount: sortedData.length,
     noContentMessage: `No matching ${which}.`,
+    sort,
```
Comment (andy7i) [inline src/pages/workspaces/workspace/workflows/WorkflowView.js:111]: is `data table` the right description for this table? (the answer might be yes, but I just wanted to double check)

File: src/components/table.js [hunk 8]
```diff
@@ -259,37 +280,52 @@ FlexTable.propTypes = {
...
+  return div({
```
Comment (andy7i) [inline src/components/table.js:321]: Why is the div better than the `h(Fragment` here? -- just curious

File: src/components/table.js (hunk 5 above)
Comment (andy7i) [reply src/components/table.js]: I think you have similar comments like this in other places. Maybe worth adding that in the other places where we offset the `colindex` and `rowindex` since folks aren't super familiar with aria rules.

File: src/components/table.js (hunk 6 above)
Comment (andy7i) [reply src/components/table.js:320]: same goes for the other 3 tables

File: src/components/table.js [hunk 9]
```diff
@@ -457,31 +554,35 @@ export const HeaderCell = props => {
 }
 
 export const Sortable = ({ sort, field, onSort, children }) => {
-  return div({
-    style: { flex: 1, display: 'flex', alignItems: 'center', cursor: 'pointer', width: '100%' },
-    onClick: () => onSort(Utils.nextSort(sort, field))
+  return h(IdContainer, [id => h(Clickable, {
```
Comment (andy7i) [inline src/components/table.js:584]: where does `id` come from?

**Review by andy7i — COMMENTED** [review]
Hey Andrew! Thanks for helping us do this important work! I appreciated the refactoring and comments you added.
I had some questions to try to understand some of the changes better (I haven't done too much front end dev recently) as well as some nitpicks. Let me know if I can help clarify anything!

File: src/components/table.js (hunk 4 above)
Comment (andrewmarcus) [reply src/components/table.js]: I am in favor of making this as easy to understand as possible! Glad to reformat. I'll add some comments too. :)

File: src/components/table.js (hunk 6 above)
Comment (andrewmarcus) [reply src/components/table.js:320]: Hmm, that's a great point. I was trying to keep these changes as unobtrusive to developers as possible, and trying to avoid having to make changes to every table.

But on the other hand, I already changed all the other tables to add support for `aria-sort`, so that's probably not a big deal at this point.

And it would certainly be better to force devs to explicitly think about accessibility where they need to. Now that I think about it, this caption is sort of like adding `alt` text to an image: sure you could just use a default but that really defeats the purpose.

Ok, I'll make it a required field, and update all of the tables accordingly.

File: src/pages/workspaces/workspace/workflows/WorkflowView.js (hunk 7 above)
Comment (andrewmarcus) [reply src/pages/workspaces/workspace/workflows/WorkflowView.js:111]: I was trying to keep the PRs isolated to either core functionality or specific pages, but I'll go ahead and make the `tableName` required, which means I'll also add accessible names to every table including this one.

File: src/components/table.js (hunk 8 above)
Comment (andrewmarcus) [reply src/components/table.js:321]: It seemed like `Fragment` doesn't actually render any markup on its own, it's just a container if you have more than one child. I changed it to a `div` since I needed to add ARIA properties and needed an actual tag to attach them to.

File: src/components/table.js (hunk 9 above)
Comment (andrewmarcus) [reply src/components/table.js:584]: The `IdContainer` generates stateful unique ids and passes them down to its children. I just copied and pasted an existing design pattern I found in use throughout the site.

I'm not entirely sure how that `id` gets passed in there, and what would happen if I had more than one child in that array; would they each be called with the `id` parameter or would only the first one be? But it seems to do what I want when written like this.

File: src/components/table.js (hunk 4 above)
Comment (zarsky-broad) [reply src/components/table.js]: Might a `Utils.cond` be simpler?

File: src/components/data/DataTable.js (hunk 1 above)
Comment (marctalbott) [reply src/components/data/DataTable.js:238]: Ah okay I think it's fine then, just wanted to double check that this wasn't meant to be a variable like it was down below

File: src/components/table.js (hunk 4 above)
Comment (andrewmarcus) [reply src/components/table.js]: This thread is outdated, I already updated it to a ternary inside some ifs. That also let me put comments in describing the various cases, which seemed useful...

File: src/pages/workspaces/workspace/workflows/WorkflowView.js (hunk 7 above)
Comment (andrewmarcus) [reply src/pages/workspaces/workspace/workflows/WorkflowView.js:111]: I confirmed with @andy7i that I'm going to keep `tableName` optional in this PR. I will publish a new PR in which I make it required and add table names to the 15+ tables which need it, in order to keep all those changes out of this PR.

File: src/components/data/DataTable.js (hunk 1 above)
Comment (andrewmarcus) [reply src/components/data/DataTable.js:238]: Yeah, it's a good question. Thanks for double checking

**Review by marctalbott — APPROVED** [review]

**Review by andy7i — APPROVED** [review]
👍 👍  Looks great! Thank you
